### PR TITLE
[ENH] Add filename template legend

### LINKED ISSUE #1198: Add legend below filename templates
Small suggestion to add a legend in the macro that generate the filename templates.

May help new comers.

It would add:

**Legend**

- entities between square brackets (for example `[_ses-<label>]`) are OPTIONAL
- for a definition of `label` and `index`, please refer to the relevant section in the [common principles](https://bids-specification.readthedocs.io/en/latest/02-common-principles.html#definitions)

### LINKED ISSUE #1220: broken JSON does not play well with pre-commit
In case anyone tries to run `pre-commit run -a` you will get a failure on this file.

https://github.com/bids-standard/bids-specification/blob/master/tools/schemacode/bidsschematools/tests/data/broken_dataset_description.json

Odds that this will be an issue with a normal use of pre-commit are low and I doubt that anyone but maintainers will run into that problem.

### PR DESCRIPTION
**Pull request by Remi-Gau**
closes #1198
fixes #1220  

### unrelated
- update install of bidsschema tool via requirements
- add black, import sort and flake8 to pre-commit config
- run black and import sort on non schema code python
- add setup.cfg with flake8 config in root folder to tell pre-commit what config to use

### CODE REVIEW

**Review by sappelhoff — COMMENTED** [review]
could you please also remove:

https://github.com/bids-standard/bids-specification/blob/d849c405eac216161a08d922c23cb4247c59c13d/.circleci/config.yml#L16

and re:

-  #1120

maybe there is a way to make the JSON pre-commit hook ignore certain files?

**Comment by Remi-Gau** [comment]
> and re:
> 
>     * #1120
> 
> 
> maybe there is a way to make the JSON pre-commit hook ignore certain files?

not for this specific hook, but will exclude the file globally in the pre-commit config
https://github.com/pre-commit/pre-commit-hooks#check-json

**Comment by Remi-Gau** [comment]
weird note: running `black .` (blacK 22.6) to lint all files reformat some files and this will be undone by the black pre-commit hook.

Most likely not ideal but probably best to investigate somewhere else than in this PR


<details> <summary>Example</summary>
```
datalad run -m 'run black on all files' 'black .'          
[INFO   ] == Command start (output follows) ===== 
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install black[jupyter]``
reformatted tools/schemacode/bidsschematools/tests/test_rules.py
reformatted tools/schemacode/bidsschematools/tests/test_render.py
reformatted tools/schemacode/bidsschematools/utils.py
reformatted tools/schemacode/bidsschematools/schema.py
reformatted tools/schemacode/bidsschematools/tests/test_validator.py
reformatted tools/schemacode/bidsschematools/render.py
reformatted tools/schemacode/bidsschematools/validator.py

All done! ✨ 🍰 ✨
7 files reformatted, 18 files left unchanged.
[INFO   ] == Command exit (modification check follows) ===== 
add(ok): tools/schemacode/bidsschematools/render.py (file)                                                                                                              
add(ok): tools/schemacode/bidsschematools/schema.py (file)                                                                                                              
add(ok): tools/schemacode/bidsschematools/tests/test_render.py (file)                                                                                                   
add(ok): tools/schemacode/bidsschematools/tests/test_rules.py (file)                                                                                                    
add(ok): tools/schemacode/bidsschematools/tests/test_validator.py (file)                                                                                                
add(ok): tools/schemacode/bidsschematools/utils.py (file)                                                                                                               
add(ok): tools/schemacode/bidsschematools/validator.py (file)                                                                                                           
Total:   0%|                                                                                                                         | 0.00/1.00 [00:00<?, ? datasets/s]CommandError: 'git -c diff.ignoreSubmodules=none commit -m '[DATALAD RUNCMD] run black on all files

=== Do not change lines below ===
{
 "chain": [],
 "cmd": "black .",
 "exit": 0,
 "extra_inputs": [],
 "inputs": [],
 "outputs": [],
 "pwd": "."
}
^^^ Do not change lines above ^^^
' --' failed with exitcode 1 under /home/remi/github/BIDS-specification
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
check yaml...........................................(no files to check)Skipped
check json...........................................(no files to check)Skipped
check python ast.........................................................Passed
check for added large files..............................................Passed
check for case conflicts.................................................Passed
black....................................................................Failed
- hook id: black
- files were modified by this hook

reformatted tools/schemacode/bidsschematools/tests/test_rules.py
reformatted tools/schemacode/bidsschematools/tests/test_render.py
reformatted tools/schemacode/bidsschematools/utils.py
reformatted tools/schemacode/bidsschematools/schema.py
reformatted tools/schemacode/bidsschematools/tests/test_validator.py
reformatted tools/schemacode/bidsschematools/validator.py
reformatted tools/schemacode/bidsschematools/render.py

All done! ✨ 🍰 ✨
7 files reformatted.

flake8...................................................................Passed
```
</details>

**Comment by Remi-Gau** [comment]
@tsalo 
when you have time could you check this weird horror that this PR is doing?

https://bids-specification--1219.org.readthedocs.build/en/1219/04-modality-specific-files/03-electroencephalography.html#eeg-recording-data

Now the filename templates include ALL datatypes everytime and the legend is rendered as codeblock. I must be missing something obvious but I can't see what.

Screenshot from 2022-08-17 11-35-27

**Comment by tsalo** [comment]
Latest looks fine to me. I'll take a look to see what might be causing problems in this PR though.

**Comment by tsalo** [comment]
I think I have the problem pinned down, but I don't have a solution yet.

The problem is that `filter_schema` checks for `dict` objects, instead of `Namespace` objects (which the schema now is). Changing `dict` below to `Mapping` addresses that problem.

https://github.com/bids-standard/bids-specification/blob/d849c405eac216161a08d922c23cb4247c59c13d/tools/schemacode/bidsschematools/schema.py#L260-L279

The next problem is that Namespace objects do not support item assignment, and `filter_schema` relies heavily on item assignment (e.g., `new_schema[k] = filtered_item`) to build the filtered version of the schema. I haven't figured out how to work around this. @effigies do you have any ideas for how to address the item assignment problem?

**Comment by Remi-Gau** [comment]
Thanks for looking into this.

Should we add a test to catch this in ci rather than having to manually check the rendering?

Also because running mkdocs serves takes "ages" to run locally which does not incentive to check.

**Comment by tsalo** [comment]
Chris had some ideas for fixing the filter that we talked about briefly on Zoom. I'll be trying them out once I finish up with a call I'm on.

>Should we add a test to catch this in ci rather than having to manually check the rendering?

We do have #788 that seems similar, but I never figured out how to actually implement it. Otherwise, I think we'll need to write unit tests for the rendering functions. I think Horea added some, but they must not be strict enough to catch this new problem.

**Comment by effigies** [comment]
I'll try to make a quick fix.

**Comment by effigies** [comment]
See https://github.com/bids-standard/bids-specification/pull/1223

File: tools/schemacode/bidsschematools/render.py
Comment (tsalo) [inline tools/schemacode/bidsschematools/render.py:310]: For some reason, this is rendered as a code block, so the Markdown elements aren't showing up.

image

Comment (effigies) [reply tools/schemacode/bidsschematools/render.py:310]: It's tabbed out.
```suggestion
    text += """

**Legend:**

- Filename entities or folders between square brackets (for example `[_ses-<label>]`) are OPTIONAL.
- See the glossary for a definition of [`label`](../99-appendices/14-glossary.md#label)
  and [`index`](../99-appendices/14-glossary.md#index).
- See the [entity page](../99-appendices/09-entities.md) for a definition of each entity.

"""
```

**Comment by Remi-Gau** [comment]
Thanks @tsalo @effigies for looking into this and fixing all of that.

@tsalo did I miss them or are the label and index definition not in the glossary?

**Comment by Remi-Gau** [comment]
> We do have #788 that seems similar, but I never figured out how to actually implement it. Otherwise, I think we'll need to write unit tests for the rendering functions. I think Horea added some, but they must not be strict enough to catch this new problem.

I may have a look into this see if I can think of something. Don't hold your breath.

File: tools/schemacode/bidsschematools/render.py
Comment (effigies) [inline tools/schemacode/bidsschematools/render.py:306]: Given this is now showing up multiple times in some pages (namely MRI), I wonder if making it collapseable would be better:

```suggestion
<details>
<summary><strong>Legend:</strong></summary>

- Filename entities or folders between square brackets (for example `[_ses-<label>]`) are OPTIONAL.
- See the glossary for a definition of [`label`](../99-appendices/14-glossary.md#label)
  and [`index`](../99-appendices/14-glossary.md#index).
- See the [entity page](../99-appendices/09-entities.md) for a definition of each entity.

</details>
```

I have obviously not tested if this would

Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py:306]: PDF rendering would have to adapted too, no?

Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py]: commit this.
let's see how the whole thing behaves

Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py]: Hum... Not quite.

Screenshot from 2022-08-18 09-57-46

Comment (tsalo) [reply tools/schemacode/bidsschematools/render.py]: It looks like the PDF doesn't show the legend at all, which I think is totally fine. I'm not sure why the legend no longer respects Markdown though.

File: tools/schemacode/bidsschematools/render.py
Comment (tsalo) [inline tools/schemacode/bidsschematools/render.py:307]: In combination with #1228, I think you can drop the second and third bullet points. I also think it would be good to have the following:

- Some entities may only allow specific values, in which case those values are listed in `<>`, separated by `|`.
- `_<suffix>` means that there are several (>6) valid suffixes for this filename pattern.
- `.<extension>` means that there are several (>6) valid extensions for this file type.
- `[.gz]` means that both the unzipped and gzipped version of the extension are valid.

File: tools/schemacode/bidsschematools/render.py
Comment (tsalo) [inline tools/schemacode/bidsschematools/render.py:302]: Once #1228 is merged, you can use `pdf_format` to toggle the admonition format. That way, if you want to include the legend in the PDF, you can change the formatting, while still keeping the admonition style in the HTML build.

File: tools/schemacode/bidsschematools/render.py
Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py:307]: The additions you suggest make sense. Will wait for those other PRs to be merged then I will add those.

Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py]: done

**Comment by Remi-Gau** [comment]
OK now the pdf looks good but not the html version.
# :shrug:

**Comment by Remi-Gau** [comment]
PDF and HTML now look good to me.

File: tools/schemacode/bidsschematools/render.py
Comment (effigies) [inline tools/schemacode/bidsschematools/render.py:493]:
```suggestion
<li><code>_&lt;suffix&gt;</code> means that there are several (>6) valid suffixes
    for this filename pattern.</li>
<li><code>.&lt;extension&gt;</code> means that there are several (>6) valid extensions
```

Comment (tsalo) [reply tools/schemacode/bidsschematools/render.py:493]: 👍 for this proposed change. The `_<suffix>` and `.<extension>` elements currently show up as `_` and `.`.

Comment (effigies) [reply tools/schemacode/bidsschematools/render.py:493]: Do we just want to do our own Markdown render? e.g., write this as:

```
- `_<suffix>` ...
```

And then something like `html = commonmark.render(md)`?

Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py:493]: > Do we just want to do our own Markdown render?

you mean as part of this PR? 
(sorry I am very dense today)

Comment (effigies) [reply tools/schemacode/bidsschematools/render.py:493]: Either way. It just looks like you're manually writing HTML, and then modifying it for the benefit of the LaTeX renderer back to something closer to Markdown. And we're hitting a case where angle brackets `<>` are hiding things in HTML when converting them to HTML escape characters is almost certainly going to mess up the TeX. I guess I'm just wondering if it's worth doing so much by hand.

Comment (Remi-Gau) [reply tools/schemacode/bidsschematools/render.py]: I used the successor to `commonmark` to do the job

File: .pre-commit-config.yaml
Comment (sappelhoff) [inline .pre-commit-config.yaml:30]: why do you prefer this over the more standard `isort`?

File: .pre-commit-config.yaml
Comment (sappelhoff) [inline .pre-commit-config.yaml:23]: should we use the most recent version?

```suggestion
    rev: '5.0.4'
```

File: setup.cfg
Comment (sappelhoff) [inline setup.cfg:5]: I think for this to work we would need https://pypi.org/project/flake8-docstrings/ ... correct?

We could add it to `requirements.txt` and then also add it to the pre-commit hook like here: https://github.com/sappelhoff/pyprep/blob/25b91e098d0055f14b5cbcae54b5ee9e0739aa62/.pre-commit-config.yaml#L27

Comment (Remi-Gau) [reply setup.cfg:5]: Hum...

I have been "wasting" some time because the linting rules of precommit and those of the ci the bidsschematools clashed, so more than happy to try to make them agree.

However I know hate this PR because it does too. Many. Things.

So I will remove all changes that are not related to the title of this PR and put them in a different one.

Comment (sappelhoff) [reply setup.cfg:5]: It's also fine with me to merge this as is and then iterate, no worries.

**Comment by Remi-Gau** [comment]
Yeah but it bothers me.

So expect a force push.

**Comment by Remi-Gau** [comment]
> Yeah but it bothers me.
> 
> So expect a force push.

OK I changed my mind: no force push. With several merge of master into this, reset and sorting what belongs to this PR or to another is going to be a pain.

So I would prefer merging and fix the pre-commit config in another PR. (pragmatism beats clean commit history on this one)

**Comment by sappelhoff** [comment]
> So I would prefer merging and fix the pre-commit config in another PR. (pragmatism beats clean commit history on this one)

yes :+1: especially because we can squash and merge it

**Review by sappelhoff — APPROVED** [review]
Just to clarify - are you ready to have this merged @Remi-Gau ?

**Comment by Remi-Gau** [comment]
> Just to clarify - are you ready to have this merged @Remi-Gau ?

yup if @tsalo and @effigies are OK with it

**Review by tsalo — APPROVED** [review]
Sorry for the delay. I'm happy with this as-is.

File: requirements.txt
Comment (effigies) [inline requirements.txt:7]: I think I suggested this, but I disagree now. This is part of the build process in RTD, and we shouldn't rely on `-e`.

```suggestion
tools/schemacode/[render]
```

File: CONTRIBUTING.md
Comment (effigies) [inline CONTRIBUTING.md:305]: So this should be restored.

File: pdf_build_src/process_markdowns.py
Comment (effigies) [inline pdf_build_src/process_markdowns.py:26]: Are there changes in this file? Style rerender?

File: setup.cfg
Comment (effigies) [reply setup.cfg:5]: `setup.cfg` is really meant for Python packages, which this isn't. If we're going to have it at the root, I would rather call it `.flake8`. See https://flake8.pycqa.org/en/latest/user/configuration.html

**Review by effigies — COMMENTED** [review]
TBH I'm finding this hard to review because of bundling process changes with semantic changes. I made a couple small comments regarding the render, which seems fine, and the contributing guide. If everything else is just process, that's fine.

**Comment by Remi-Gau** [comment]
@effigies 
I can close this and start a cleaner PR, if you prefer.
This thing has become a headache to review and to work with TBF. Sorry. :-(

**Comment by effigies** [comment]
Sorry, I hate to be annoying, but I really would prefer to keep these things separate. If it would save you trouble, I'd be happy to unsentimentally squash what I think are the relevant changes into a single commit with you as author, and you can just review that.

**Comment by Remi-Gau** [comment]
> Sorry, I hate to be annoying, but I really would prefer to keep these things separate. If it would save you trouble, I'd be happy to unsentimentally squash what I think are the relevant changes into a single commit with you as author, and you can just review that.

No worries. I was getting annoyed at how this PR was started to mix things that were unrelated but I think I gave up too quickly to sort it.

Will close and open separate PRs.

**Pull request closed without merging**
